Add Inductor mm template in TLX

dshi7 opened the pull request:
Summary:
Adds new Triton TLX mm templates. The changes include:

- New TLX mm template (gemm-CLC)
- Template registration: Added templates in mm.py
- Config heuristics: Added and registered config heuristic configs in torch/_inductor/fb/tlx_templates/registry.py
- Integration: Modified tuned_mm() to include the TLX template as a choice when conditions are met
- Tests: Added test_blackwell_tlx_mm test case

NOTE: We plan to gate most future iterations (modifying templates and tuning heuristics/config options) within the `fb/` folder until they are production-ready.

Differential Revision: D89151760




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

dshi7 commented:
@pytorchbot label "topic: not user facing"

eellison reviewed (commented):
looks generally reasonable, maybe cc @drisspg , @njriasan @PaulZhang12 for thoughts

dshi7 commented on torch/_inductor/kernel/mm.py:
```diff
@@ -423,6 +425,10 @@ def _to_dtype(x):
             if use_triton_blackwell_tma_template(mat1, mat2, output_layout=layout):
                 templates_to_use.append(blackwell_ws_persistent_device_tma_mm_template)
 
+            if inductor_config.enable_tlx():
```
can you provide an example for me to follow?

drisspg commented on test/inductor/test_max_autotune_blackwell.py:
```diff
@@ -0,0 +1,365 @@
+            read_api
+        ).check_count(write_api, write_count).run(code[0])
+
+        torch.testing.assert_close(c_actual, c_expected, atol=1e-2, rtol=1e-2)
```
Are these tolerances right for fp16?

dshi7 replied:
I was following existing examples. should I leave it blank to use the default precision?

```
_DTYPE_PRECISIONS = {
    torch.float16: (0.001, 1e-5),
    torch.bfloat16: (0.016, 1e-5),
    torch.float32: (1.3e-6, 1e-5),
    torch.float64: (1e-7, 1e-7),
    torch.complex32: (0.001, 1e-5),
    torch.complex64: (1.3e-6, 1e-5),
    torch.complex128: (1e-7, 1e-7),
}
```

PaulZhang12 commented on torch/_inductor/template_heuristics/triton.py:
```diff
@@ -2139,6 +2149,58 @@ def __init__(self) -> None:
+
+if inductor_config.enable_tlx():
+    from torch._inductor.fb.tlx_templates.mm_templates import tlx_blackwell_gemm_clc_template, tlx_blackwell_gemm_2cta_template
+    class BlackwellTLXTemplateConfigMixin(TMATemplateConfigMixin):
```
Yeah this we might want to do outside of PyTorch OSS. We should be able to register something like this, maybe similar to DiodeInductorChoices?

dshi7 replied:
Both registry and referenced classes are defined in the same file, so registering TLX in a separate file will cause import cycling. I can do it but wonder if we really  need to do it? @eellison @PaulZhang12

eellison reviewed (approved):
Looks good to me, thank you ! the internal registration will also be easier for you to iterate faster on without oss tests, so it's better all around I think.

cc @drisspg for final comments

drisspg reviewed (approved):
Nice, super clean

dshi7 commented on test/inductor/test_max_autotune_blackwell.py:
```diff
@@ -291,6 +291,79 @@ def addmm(x, a, b):
         torch.testing.assert_close(c_actual, c_expected, atol=1e-2, rtol=1e-2)
 
+    @unittest.skipIf(
+        not has_datacenter_blackwell_tma_device() or not config.is_fbcode(),
```
I'm not familiar with how OSS CI works. Adding it just because this test is also exposed to OSS.

dshi7 commented on test/inductor/test_max_autotune_blackwell.py:
```diff
@@ -291,6 +291,79 @@ def addmm(x, a, b):
+        dynamic: bool,
+        epilogue_subtile: bool,
+    ):
+        try:
```
I will remove this try-block after https://fb.workplace.com/groups/420659799592399/permalink/1185832739741764/

dshi7 committed:
[Inductor] Add TLX mm templates (#170410)

Summary:

Adds new Triton TLX mm templates. The changes include:

- New TLX mm template (gemm-CLC)
- Template registration: Added templates in mm.py
- Config heuristics: Added and registered config heuristic configs in template_heuristics/triton.py
- Integration: Modified tuned_mm() to include the TLX template as a choice when conditions are met
- Tests: Added test_blackwell_tlx_mm test case

NOTE: We plan to gate most future iterations (modifying templates and tuning heuristics/config options) within the `fb/` folder until they are production-ready.

Reviewed By: htyu, njriasan

Differential Revision: D89151760
